File: apps/plugin-websearch/Sources/TobyPluginWebSearchLib/Prompts.swift
import Foundation

public enum Prompts {
	static let systemPromptSection = """
	### Web Search
	You can search the web using the webSearch tool (powered by Brave Search). Use it when the user asks about current events, facts, research, or anything that requires up-to-date information from the web. Always cite source URLs from search results.
	"""

	static let singleSessionRules = """
	You are assisting with web search. Use the webSearch tool to find information when the user asks questions that require looking up facts, current events, or research. Summarize and synthesize search results clearly, citing source URLs when relevant.
	"""

	static let singleSessionUserTemplate = """
	User request (web search):
	{{userPrompt}}
	"""

	static let multiUserContentTemplate = """
	## Web Search context
	The user may want to search the web. Use webSearch when current information or research is needed.

	User request (may also mention other integrations):
	{{userPrompt}}
	"""

	public static func buildChatModelPrep() -> [String: Any] {
		[
			"systemPromptSection": systemPromptSection,
			"singleSessionRules": singleSessionRules,
			"singleSessionUserTemplate": singleSessionUserTemplate,
			"multiUserContentTemplate": multiUserContentTemplate,
		]
	}

	public static func buildChatReadiness(config: [String: Any], state: [String: Any]) -> [String: Any] {
		if PluginOutput.isConnected(config: config, state: state) || SearchClient.hasApiKey(config: config) {
			return ["ok": true]
		}
		return [
			"ok": false,
			"hint":
				"Add a Brave Search API key in `toby configure` under Web Search, or run `toby connect websearch`.",
		]
	}
}

File: apps/plugin-news/src/prompts.ts
type JsonRecord = Record<string, unknown>;

export function buildChatModelPrep(): JsonRecord {
	return {
		systemPromptSection: `### News
You can fetch latest headlines and search recent articles via The Guardian Open Platform. Always attribute stories to The Guardian and include article URLs. Never invent headlines.`,
		singleSessionRules: `You are a news assistant. Headlines and article metadata come from The Guardian's free Open Platform API.

Tools:
- **getLatestNews** — Latest headlines. Optional section (world, us-news, uk-news, technology, business, sport, science, environment, culture, politics) and limit.
- **searchNews** — Search recent articles by topic, person, place, or event. Requires query.
- **askUser** — For user choices; collect answers only through this tool.

Rules:
- Never invent news. Only report articles returned by the tools.
- Prefer **searchNews** when the user names a subject; use **getLatestNews** for general "what's in the news" requests.
- Always name **The Guardian** as the source and include each article URL.
- Include publication dates when present.
- If the API key is missing or invalid, tell the user to add a free Guardian Open Platform key in Toby Integrations → News.
- Treat this as a headline briefing, not a full-text archive. Offer the URL when the user wants the complete article.`,
		singleSessionUserTemplate: `User request (News):
{{userPrompt}}`,
		multiUserContentTemplate: `## News
Use News tools for headlines and recent article search. Attribute The Guardian and include URLs.

If you need a decision from the user, call **askUser** with options.

User request (may also mention other integrations):
{{userPrompt}}`,
	};
}

export function buildChatReadiness(
	config: JsonRecord,
	state: JsonRecord,
): JsonRecord {
	if (state.connectedAt || String(config.apiKey ?? "").trim()) {
		return { ok: true };
	}
	return {
		ok: false,
		hint: "Add a free Guardian Open Platform API key in `toby configure`, then run `toby connect news`.",
	};
}
